pkg/ffmpeg/codec_detector.go

```diff
@@ -0,0 +1,80 @@
+package ffmpeg
+
+import (
+	"context"
+	"fmt"
+)
+
+var HWCodecSupport []StreamFormat
+
+/*
+Tests all hardware codec's to see if they work
+*/
+func FindHWCodecs(ctx context.Context, encoder FFMpeg) {
+	for _, codec := range []StreamFormat{
+		StreamFormatN264,
+		StreamFormatI264,
+		/*
+			Untested:
+				StreamFormatA264,
+				StreamFormatM264,
+				StreamFormatV264,
+				StreamFormatO264,
+				StreamFormatIVP9,
+				StreamFormatVVP9,
+		*/
+	} {
+		var args Args
+		args = append(args, "-hide_banner")
```

Comment:
For vaapi hw accelarated encoding an extra step is needed to select the device (for most/simple cases /dev/dri/renderD128 should work).
Also one more filter to convert to a vaapi supported format and upload to the device
For example a minimal working ffmpeg command (usig an old Vega56 amd cardf would be 
```
ffmpeg -vaapi_device /dev/dri/renderD128 -f lavfi -i color=c=red -t 1 -vf format=nv12,hwupload -c:v h264_vaapi test.mp4
```
The above should work for detection purposes, for the encoding i am not sure if the issue mentioned here <https://github.com/stashapp/stash/pull/1041> (amd specific) is still valid. I will post again after testing properly.

Reply:
I have pushed a commit which preps appropriate hw devices such as vaapi_device, it also attempts to use hw scaling.
I can't test AMD, but it works for NVidia and Intel.
It also contains code for full hardware transcoding, which remains unused since i dont have a good strategy for how i should use it.

Reply:
After testing a bit ( Vega56 Card using VAAPI via the opensource mesa driver ) it seems that the issue in #1041  ( empty_moov movflag option for h264_vaapi) is still present. Not sure if it is specific to my card or for all amd cards.
Given the above i couldnt test properly. A few issues (VAAPI related) with the latest commits

1.  The way HWCodecPrepend_Encode is used you end up with two different filter sections when adding the scale filter
2. In HWCodecFilter appending the `hwdownload` is not needed and breaks the encoding. Also due to 1 you already have a video filter with the format and hwupload so the end result is also broken
3. The StreamFormatV264 needed some tweaking

The closest i get it to work was with the below
```diff
diff --git a/pkg/ffmpeg/codec_detector.go b/pkg/ffmpeg/codec_detector.go
index a2cc63e8..c6c129db 100644
--- a/pkg/ffmpeg/codec_detector.go
+++ b/pkg/ffmpeg/codec_detector.go
@@ -14,6 +14,7 @@ func FindHWCodecs(ctx context.Context, encoder FFMpeg) {
 	for _, codec := range []StreamFormat{
 		StreamFormatN264,
 		StreamFormatI264,
+		StreamFormatV264,
 		/*
 			Untested:
 				StreamFormatA264,
@@ -32,14 +33,12 @@ func FindHWCodecs(ctx context.Context, encoder FFMpeg) {
 		args = args.Input("color=c=red")
 		args = args.Duration(0.1)
 
-		args = HWCodecPrepend_Encode(args, codec.codec)
 		args = args.VideoCodec(codec.codec)
 		if len(codec.extraArgs) > 0 {
 			args = append(args, codec.extraArgs...)
 		}
-
+		videoFilter := HWFilterInit(codec.codec)
 		// Test scaling
-		var videoFilter VideoFilter
 		videoFilter = videoFilter.ScaleDimensions(-2, 160)
 		videoFilter = HWCodecFilter(videoFilter, codec.codec)
 		args = args.VideoFilter(videoFilter)
@@ -93,7 +92,7 @@ func HWCodecVideoSupported(ctx context.Context, encoder FFMpeg, o TranscodeStrea
 	args = args.Duration(0.1)
 
 	// Test scaling
-	var videoFilter VideoFilter
+	videoFilter := HWFilterInit(o.Codec.codec)
 	videoFilter = videoFilter.ScaleDimensions(-2, 160)
 	videoFilter = HWCodecFilter(videoFilter, o.Codec.codec)
 	args = args.VideoFilter(videoFilter)
@@ -120,23 +119,24 @@ func HWCodecDevice_Encode(args Args, codec VideoCodec) Args {
 		args = append(args, "0")
 	case VideoCodecV264,
 		VideoCodecVVP9:
-		args = append(args, "-hwaccel_device")
+		args = append(args, "-vaapi_device")
 		args = append(args, "/dev/dri/renderD128")
 	}
 
 	return args
 }
 
-// Prepend codec for hardware encoding only
-func HWCodecPrepend_Encode(args Args, codec VideoCodec) Args {
+// Initialise a video filter for HW encoding
+func HWFilterInit(codec VideoCodec) VideoFilter {
+	var videoFilter VideoFilter
 	switch codec {
 	case VideoCodecV264,
 		VideoCodecVVP9:
-		args = append(args, "-vf")
-		args = append(args, "format=nv12,hwupload")
+		videoFilter = videoFilter.Append("format=nv12")
+		videoFilter = videoFilter.Append("hwupload")
 	}
 
-	return args
+	return videoFilter
 }
 
 /*
@@ -156,11 +156,7 @@ func HWCodecDevice_Full(args Args, codec VideoCodec) Args {
 		args = append(args, "0")
 	case VideoCodecV264,
 		VideoCodecVVP9:
-		args = append(args, "-hwaccel")
-		args = append(args, "vaapi")
-		args = append(args, "-hwaccel_output_format")
-		args = append(args, "vaapi")
-		args = append(args, "-hwaccel_device")
+		args = append(args, "-vaapi_device")
 		args = append(args, "/dev/dri/renderD128")
 	case VideoCodecI264,
 		VideoCodecIVP9:
@@ -183,7 +179,7 @@ func HWCodecFilter(args VideoFilter, codec VideoCodec) VideoFilter {
 			args = VideoFilter(strings.Replace(sargs, "scale=", "hwupload_cuda,scale_cuda=", 1)).Append("hwdownload")
 		case VideoCodecV264,
 			VideoCodecVVP9:
-			args = VideoFilter(strings.Replace(sargs, "scale=", "hwupload,scale_vaapi=", 1)).Append("hwdownload")
+			args = VideoFilter(strings.Replace(sargs, "scale=", "scale_vaapi=", 1))
 			// BUG: scale_qsv is seemingly broken on windows?
 			/*case VideoCodecI264,
 			VideoCodecIVP9:
diff --git a/pkg/ffmpeg/stream.go b/pkg/ffmpeg/stream.go
index 24d754ef..c6d0b04b 100644
--- a/pkg/ffmpeg/stream.go
+++ b/pkg/ffmpeg/stream.go
@@ -140,8 +140,7 @@ var (
 		MimeType: MimeMp4,
 		extraArgs: []string{
 			"-movflags", "frag_keyframe+empty_moov",
-			"-quality", "50",
-			"-pix_fmt", "yuv420p",
+			"-qp", "20",
 		},
 	}
 
@@ -153,7 +152,7 @@ var (
 		format:   FormatMP4,
 		MimeType: MimeMp4,
 		extraArgs: []string{
-			"-movflags", "frag_keyframe",
+			"-movflags", "frag_keyframe+empty_moov",
 			"-pix_fmt", "yuv420p",
 		},
 	}
@@ -289,7 +288,7 @@ func (o TranscodeStreamOptions) getStreamArgs() Args {
 		args = args.Duration(hlsSegmentLength)
 	}
 
-	args = HWCodecPrepend_Encode(args, o.Codec.codec)
+	videoFilter := HWFilterInit(o.Codec.codec)
 	args = args.Input(o.Input)
 
 	if o.VideoOnly {
@@ -300,7 +299,6 @@ func (o TranscodeStreamOptions) getStreamArgs() Args {
 
 	// don't set scale when copying video stream
 	if o.Codec.codec != VideoCodecCopy {
-		var videoFilter VideoFilter
 		videoFilter = videoFilter.ScaleMax(o.VideoWidth, o.VideoHeight, o.MaxTranscodeSize)
 		videoFilter = HWCodecFilter(videoFilter, o.Codec.codec)
 		args = args.VideoFilter(videoFilter)
```

Reply:
I have some concerns, one is that pixel format isn't enforced, which may cause browser incompatibilities.
The reason hwupload and hwdownload was there was specifically to accommodate changing the format, but it might be the case that the encoder is nice enough to be consistent (NVENC isnt).
I also assume you mean to always apply the video filter, so i moved it out of the scale check (in stream.go).
https://github.com/stashapp/stash/blob/6295dc4d5d70e762959b4e1d88fc944f39f8b5cf/pkg/ffmpeg/stream.go#L300-L305

Reply:
I changed the intel and nvidia encoder to this format.
Given that i can't really test the vaapi encoder, i think it should be disabled for this commit, unless you want to work on it.
So it can be pushed ahead, and maybe someone with an actual amd gpu can make a commit fixing the amd encoders.